File: FwiCore/com/fiision/lib/network/request/FwiRequest.m
#import "FwiRequest.h"
#import "FwiDataParam.h"
#import "FwiFormParam.h"
#import "FwiMultipartParam.h"


#define kHTTP_Copy      @"COPY";
#define kHTTP_Delete    @"DELETE";
#define kHTTP_Get       @"GET";
#define kHTTP_Link      @"LINK";
#define kHTTP_Head      @"HEAD";
#define kHTTP_Options   @"OPTIONS";
#define kHTTP_Patch     @"PATCH";
#define kHTTP_Post      @"POST";
#define kHTTP_Purge     @"PURGE";
#define kHTTP_Put       @"PUT";
#define kHTTP_Unlink    @"UNLINK";


static inline NSString* FwiGenerateUserAgent() {
    __autoreleasing NSDictionary *bundleInfo   = [[NSBundle mainBundle] infoDictionary];
    __autoreleasing UIDevice *deviceInfo       = [UIDevice currentDevice];
    __autoreleasing NSString *bundleExecutable = [bundleInfo objectForKey:(NSString *)kCFBundleExecutableKey];
    __autoreleasing NSString *bundleIdentifier = [bundleInfo objectForKey:(NSString *)kCFBundleIdentifierKey];
    __autoreleasing NSString *bundleVersion    = [bundleInfo objectForKey:(NSString *)kCFBundleVersionKey];
    __autoreleasing NSString *systemVersion    = [deviceInfo systemVersion];
    __autoreleasing NSString *model            = [deviceInfo model];
    
    // Define user-agent
    return [NSString stringWithFormat:@"%@/%@ (%@; iOS %@; Scale/%0.2f)", (bundleExecutable ? bundleExecutable : bundleIdentifier), bundleVersion, model, systemVersion, [[UIScreen mainScreen] scale]];
}


@interface FwiRequest () {

    // Request type
    FwiHttpMethod    _type;

    // Raw request
    FwiDataParam *_raw;
    // Form request
    NSMutableArray   *_form;
    NSMutableArray   *_upload;
}

@property (nonatomic, strong) FwiDataParam *raw;
@property (nonatomic, strong) NSMutableArray *form;
@property (nonatomic, strong) NSMutableArray *upload;


/** Class's constructors. */
- (id)initWithURL:(NSURL *)url methodType:(FwiHttpMethod)type;


/** Initialize form & upload. */
- (void)_initForm;
- (void)_initUpload;

/** Add form parameter. */
- (void)_addFormParam:(FwiFormParam *)param;

/** Add multipart parameter. */
- (void)_addMultipartParam:(FwiMultipartParam *)param;

@end


@implementation FwiRequest


#pragma mark - Class's constructors
- (id)initWithURL:(NSURL *)url methodType:(FwiHttpMethod)type {
    self = [super initWithURL:url cachePolicy:NSURLRequestReloadIgnoringCacheData timeoutInterval:60.0f];
    if (self) {
        _type = type;
        _raw = nil;
        _form = nil;
        _upload = nil;
        
        // Assign HTTP Method
        switch (_type) {
            case kCopy: {
                self.HTTPMethod = kHTTP_Copy;
                break;
            }
            case kDelete: {
                self.HTTPMethod = kHTTP_Delete;
                break;
            }
            case kHead: {
                self.HTTPMethod = kHTTP_Head;
                break;
            }
            case kLink: {
                self.HTTPMethod = kHTTP_Link;
                break;
            }
            case kOptions: {
                self.HTTPMethod = kHTTP_Options;
                break;
            }
            case kPatch: {
                self.HTTPMethod = kHTTP_Patch;
                break;
            }
            case kPost: {
                self.HTTPMethod = kHTTP_Post;
                break;
            }
            case kPurge: {
                self.HTTPMethod = kHTTP_Purge;
                break;
            }
            case kPut: {
                self.HTTPMethod = kHTTP_Put;
                break;
            }
            case kUnlink: {
                self.HTTPMethod = kHTTP_Unlink;
                break;
            }
            case kGet:
            default: {
                self.HTTPMethod = kHTTP_Get;
                break;
            }
        }
    }
    return self;
}


#pragma mark - Cleanup memory
- (void)dealloc {
    self.raw    = nil;
    self.form   = nil;
    self.upload = nil;

#if !__has_feature(objc_arc)
    [super dealloc];
#endif
}


#pragma mark - Class's properties


#pragma mark - Class's public methods
- (size_t)prepare {
    __autoreleasing NSDictionary *allHeaders = [self allHTTPHeaderFields];
    if (!allHeaders[@"Accept"]) [self setValue:@"*/*" forHTTPHeaderField:@"Accept"];
    if (!allHeaders[@"Accept-Encoding"]) [self setValue:@"gzip" forHTTPHeaderField:@"Accept-Encoding"];
    if (!allHeaders[@"Accept-Charset"]) [self setValue:@"UTF-8" forHTTPHeaderField:@"Accept-Charset"];
    if (!allHeaders[@"Connection"]) [self setValue:@"close" forHTTPHeaderField:@"Connection"];
    if (!allHeaders[@"User-Agent"]) [self setValue:FwiGenerateUserAgent() forHTTPHeaderField:@"User-Agent"];

    /* Condition validation */
    if (!_raw && _form.count == 0 && _upload.count == 0) return 0;
    size_t length = 0;

    if (_raw) {
        length = [_raw.data length];
        [self setHTTPBody:_raw.data];
        [self setValue:_raw.contentType forHTTPHeaderField:@"Content-Type"];
        [self setValue:[NSString stringWithFormat:@"%zu", length] forHTTPHeaderField:@"Content-Length"];
    }
    else {
        switch (_type) {
            case kPatch:
            case kPost:
            case kPut: {
                if (_form && !_upload) {
                    __autoreleasing NSData *encodedData = [[_form componentsJoinedByString:@"&"] toData];

                    length = [encodedData length];
                    [self setHTTPBody:encodedData];
                    [self setValue:@"application/x-www-form-urlencoded" forHTTPHeaderField:@"Content-Type"];
                    [self setValue:[NSString stringWithFormat:@"%zu", length] forHTTPHeaderField:@"Content-Length"];
                }
                else {
                    // Define boundary
                    __unsafe_unretained __block NSString *boundary = [NSString stringWithFormat:@"--------%li", (unsigned long) [[NSDate date] timeIntervalSince1970]];
                    __autoreleasing NSString *contentType = [NSString stringWithFormat:@"multipart/form-data; boundary=%@", boundary];

                    // Define body
                    __unsafe_unretained __block NSMutableData *body = [NSMutableData data];
                    
                    [_form enumerateObjectsUsingBlock:^(FwiFormParam *pair, NSUInteger idx __attribute__((unused)), BOOL *stop __attribute__((unused))) {
                        [body appendData:[[NSString stringWithFormat:@"\r\n--%@\r\n", boundary] toData]];
                        [body appendData:[[NSString stringWithFormat:@"Content-Disposition: form-data; name=\"%@\"\r\n\r\n", pair.key] toData]];
                        [body appendData:[pair.value toData]];
                    }];
                    [_upload enumerateObjectsUsingBlock:^(FwiMultipartParam *part, NSUInteger idx __attribute__((unused)), BOOL *stop __attribute__((unused))) {
                        [body appendData:[[NSString stringWithFormat:@"\r\n--%@\r\n", boundary] toData]];
                        [body appendData:[[NSString stringWithFormat:@"Content-Disposition: form-data; name=\"%@\"; filename=\"%@\"\r\n", part.name, part.fileName] toData]];
                        [body appendData:[[NSString stringWithFormat:@"Content-Type: %@\r\n\r\n", part.contentType] toData]];
                        [body appendData:part.data];
                    }];
                    [body appendData:[[NSString stringWithFormat:@"\r\n--%@--\r\n", boundary] toData]];

                    length = [body length];
                    [self setHTTPBody:body];
                    [self setValue:contentType forHTTPHeaderField:@"Content-Type"];
                    [self setValue:[NSString stringWithFormat:@"%zu", length] forHTTPHeaderField:@"Content-Length"];
                }
                break;
            }
            case kDelete:
            case kGet:
            default: {
                __autoreleasing NSString *finalURL = [NSString stringWithFormat:@"%@?%@", self.URL.absoluteString, [_form componentsJoinedByString:@"&"]];
                self.URL = [NSURL URLWithString:finalURL];
                break;
            }
        }
    }
    return length;
}


#pragma mark - Class's private methods
- (void)_initForm {
    /* Condition validation */
    if (_form) return;
    _form = [[NSMutableArray alloc] initWithCapacity:9];
}
- (void)_initUpload {
    /* Condition validation */
    if (_upload) return;
    _upload = [[NSMutableArray alloc] initWithCapacity:1];
}

- (void)_addFormParam:(FwiFormParam *)param {
    /* Condition validation */
    if (![param isMemberOfClass:[FwiFormParam class]] || [_form containsObject:param]) return;
    [_form addObject:param];
}

- (void)_addMultipartParam:(FwiMultipartParam *)param {
    /* Condition validation */
    if (![param isMemberOfClass:[FwiMultipartParam class]] || [_upload containsObject:param]) return;
    [_upload addObject:param];
}


@end


@implementation FwiRequest (FwiRequestCreation)


#pragma mark - Class's static constructors
+ (__autoreleasing FwiRequest *)requestWithURL:(NSURL *)url methodType:(FwiHttpMethod)type {
    return FwiAutoRelease([[FwiRequest alloc] initWithURL:url methodType:type]);
}


@end


@implementation FwiRequest (FwiForm)


- (void)setFormParam:(FwiFormParam *)param {
    if (!_form) [self _initForm];
    else [_form removeAllObjects];
    FwiRelease(_raw);
    
    [self _addFormParam:param];
}
- (void)setFormParams:(NSArray *)params {
    if (!_form) [self _initForm];
    else [_form removeAllObjects];
    FwiRelease(_raw);
    
    [params enumerateObjectsUsingBlock:^(FwiFormParam *param, NSUInteger idx __attribute__((unused)), BOOL *stop __attribute__((unused))) {
        [self _addFormParam:param];
    }];
}
- (void)addFormParam:(FwiFormParam *)param {
    if (!_form) [self _initForm];
    FwiRelease(_raw);
    
    [self _addFormParam:param];
}
- (void)addFormParams:(NSArray *)params {
    if (!_form) [self _initForm];
    FwiRelease(_raw);
    
    [params enumerateObjectsUsingBlock:^(FwiFormParam *param, NSUInteger idx __attribute__((unused)), BOOL *stop __attribute__((unused))) {
        [self _addFormParam:param];
    }];
}

- (void)setMultipartParam:(FwiMultipartParam *)param {
    if (!_upload) [self _initUpload];
    else [_upload removeAllObjects];
    FwiRelease(_raw);
    
    [self _addMultipartParam:param];
}
- (void)setMultipartParams:(NSArray *)params {
    if (!_upload) [self _initUpload];
    else [_upload removeAllObjects];
    FwiRelease(_raw);
    
    [params enumerateObjectsUsingBlock:^(FwiMultipartParam *param, NSUInteger idx __attribute__((unused)), BOOL *stop __attribute__((unused))) {
        [self _addMultipartParam:param];
    }];
}
- (void)addMultipartParam:(FwiMultipartParam *)param {
    if (!_upload) [self _initUpload];
    FwiRelease(_raw);
    
    [self _addMultipartParam:param];
}
- (void)addMultipartParams:(NSArray *)params {
    if (!_upload) [self _initUpload];
    FwiRelease(_raw);
    
    [params enumerateObjectsUsingBlock:^(FwiMultipartParam *param, NSUInteger idx __attribute__((unused)), BOOL *stop __attribute__((unused))) {
        [self _addMultipartParam:param];
    }];
}


@end


@implementation FwiRequest (FwiRaw)


- (void)setDataParameter:(FwiDataParam *)parameter {
    /* Condition validation: Validate method type */
    if (!(_type == kPost || _type == kPatch || _type == kPut) || !parameter) return;

    // Release form
    FwiRelease(_form);
    FwiRelease(_upload);

    // Keep new raw
    FwiRelease(_raw);
    _raw = FwiRetain(parameter);
}


@end

File: FwiCore/com/fiision/lib/network/FwiRequest.swift
//  Project name: FwiCore
//  File name   : FwiRequest.swift
//
//  Author      : Phuc, Tran Huu
//  Created date: 12/3/14
//  Version     : 1.00
//  --------------------------------------------------------------
//  Copyright © 2012, 2016 Fiision Studio.
//  All Rights Reserved.
//  --------------------------------------------------------------
//
//  Permission is hereby granted, free of charge, to any person obtaining  a  copy
//  of this software and associated documentation files (the "Software"), to  deal
//  in the Software without restriction, including without limitation  the  rights
//  to use, copy, modify, merge,  publish,  distribute,  sublicense,  and/or  sell
//  copies of the Software,  and  to  permit  persons  to  whom  the  Software  is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF  ANY  KIND,  EXPRESS  OR
//  IMPLIED, INCLUDING BUT NOT  LIMITED  TO  THE  WARRANTIES  OF  MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO  EVENT  SHALL  THE
//  AUTHORS OR COPYRIGHT HOLDERS  BE  LIABLE  FOR  ANY  CLAIM,  DAMAGES  OR  OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING  FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN  THE
//  SOFTWARE.
//
//
//  Disclaimer
//  __________
//  Although reasonable care has been taken to  ensure  the  correctness  of  this
//  software, this software should never be used in any application without proper
//  testing. Fiision Studio disclaim  all  liability  and  responsibility  to  any
//  person or entity with respect to any loss or damage caused, or alleged  to  be
//  caused, directly or indirectly, by the use of this software.

import UIKit
import Foundation


public final class FwiRequest: NSMutableURLRequest {

    // MARK: Class's properties
    fileprivate var methodType: FwiHttpMethod?

    fileprivate var raw: FwiDataParam?
    fileprivate var form: [FwiFormParam]?
    fileprivate var upload: [FwiMultipartParam]?

    // MARK: Class's public methods
    @discardableResult
    public func prepare() -> UInt {
        // Predefine headers
        self.defineUserAgent()
        if  value(forHTTPHeaderField: "Accept") == nil {
            setValue("*/*", forHTTPHeaderField: "Accept")
        }
        if value(forHTTPHeaderField: "Accept-Charset") == nil {
            setValue("UTF-8", forHTTPHeaderField: "Accept-Charset")
        }
        if value(forHTTPHeaderField: "Connection") == nil {
            setValue("close", forHTTPHeaderField: "Connection")
        }

        /* Condition validation */
        if raw == nil && form == nil && upload == nil {
            return 0
        }

        /* Condition validation: validate method type */
        guard let type = methodType else {
            return 0
        }

        guard let r = raw else {
            switch type {
            case .get:
                let query = form?.map {$0.description}.joined(separator: "&")
                if let url = url?.absoluteString, let q = query {
                    self.url = URL(string: "\(url)?\(q)")
                }
                return 0

            case .patch, .post, .put:
                if form != nil && upload == nil {
                    setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")

                    let query = form?.map {$0.description}.joined(separator: "&")
                    if let data = query?.toData() {
                        setValue("\(data.count)", forHTTPHeaderField: "Content-Length")
                        httpBody = data
                        return UInt(data.count)
                    }
                    return 0

                } else {
                    let body = NSMutableData()
                    let boundary = "----------\(Date().timeIntervalSince1970)"
                    setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

                    // Multi files
                    upload?.forEach({
                        if let
                            boundaryData = "\r\n--\(boundary)\r\n".toData(),
                            let contentType = "Content-Type: \($0.contentType)\r\n\r\n".toData(),
                            let contentDisposition = "Content-Disposition: form-data; name=\"\($0.name)\"; filename=\"\($0.fileName)\"\r\n".toData() {

                            body.append(boundaryData)
                            body.append(contentDisposition)
                            body.append(contentType)
                            body.append($0.contentData as Data)
                        }
                    })

                    // Multi params
                    form?.forEach({
                        if let
                            contentData = $0.value.toData(),
                            let boundaryData = "\r\n--\(boundary)\r\n".toData(),
                            let contentDisposition = "Content-Disposition: form-data; name=\"\($0.key)\"\r\n\r\n".toData() {

                            body.append(boundaryData)
                            body.append(contentDisposition)
                            body.append(contentData)
                        }
                    })

                    // Finalize request
                    if let boundaryData = "\r\n--\(boundary)\r\n".toData() {
                        body.append(boundaryData)
                    }

                    setValue("\(body.length)", forHTTPHeaderField: "Content-Length")
                    httpBody = body as Data

                    return UInt(body.length)
                }

            default:
                return 0
            }
        }

        // Handle raw data request
        httpBody = r.data as Data
        setValue(r.contentType, forHTTPHeaderField: "Content-Type")
        setValue("\(r.data.count)", forHTTPHeaderField: "Content-Length")

        return UInt(r.data.count)
    }

    /** Set custom data param. */
    public func setDataParam(_ param: FwiDataParam?) {
        /* Condition validation: Validate method type */
        if !(methodType == .post || methodType == .patch || methodType == .put) {
            return
        }
        raw = param
        form = nil
        upload = nil
    }

    /** Add form parameter. */
    public func addFormParam(_ param: FwiFormParam?) {
        self.initializeForm()
        raw = nil

        if let paramsForm = param {
            form?.append(paramsForm)
        }
    }
    public func setFormParam(_ param: FwiFormParam?) {
        form?.removeAll(keepingCapacity: false)
        self.addFormParam(param)
    }
    public func addFormParams(_ params: [FwiFormParam]?) {
        self.initializeForm()
        raw = nil

        params?.forEach() { [weak self] in
            self?.form?.append($0)
        }
    }
    public func setFormParams(_ params: [FwiFormParam]?) {
        form?.removeAll(keepingCapacity: false)
        self.addFormParams(params)
    }

    /** Add multipart parameter. */
    public func addMultipartParam(_ param: FwiMultipartParam?) {
        self.initializeUpload()
        raw = nil

        if let p = param {
            upload?.append(p)
        }
    }
    public func setMultipartParam(_ param: FwiMultipartParam?) {
        upload?.removeAll(keepingCapacity: false)
        self.addMultipartParam(param)
    }
    public func addMultipartParams(_ params: [FwiMultipartParam]?) {
        self.initializeUpload()
        raw = nil

        params?.forEach() { [weak self] in
            self?.upload?.append($0)
        }
    }
    public func setMultipartParams(_ params: [FwiMultipartParam]?) {
        upload?.removeAll(keepingCapacity: false)
        self.addMultipartParams(params)
    }

    // MARK: Class's private methods
    fileprivate func initializeForm() {
        if form != nil {
            return
        }
        form = [FwiFormParam]()
    }
    fileprivate func initializeUpload() {
        if upload != nil {
            return
        }
        upload = [FwiMultipartParam]()
    }

    /** Define user agent for each request. */
    fileprivate func defineUserAgent() {
        let deviceInfo = UIDevice.current
        let bundleInfo = Bundle.main.infoDictionary
        let bundleVersion = (bundleInfo?[kCFBundleVersionKey as String] as? String) ?? ""
        let bundleIdentifier = (bundleInfo?[kCFBundleIdentifierKey as String] as? String) ?? ""

        let userAgent = "\(bundleIdentifier)/\(bundleVersion) (\(deviceInfo.model); iOS \(deviceInfo.systemVersion); Scale/\(UIScreen.main.scale))"
        setValue(userAgent, forHTTPHeaderField: "User-Agent")
    }
}


// Creation
public extension FwiRequest {

    // MARK: Class's constructors
    public convenience init(url: URL, httpMethod method: FwiHttpMethod = .get) {
        self.init(url: url, cachePolicy: NSURLRequest.CachePolicy.reloadIgnoringLocalCacheData, timeoutInterval: 30.0)
        raw = nil
        form = nil
        upload = nil
        methodType = method

        switch method {
        case .copy:
            self.httpMethod = "COPY"
            break

        case .delete:
            self.httpMethod = "DELETE"
            break

        case .head:
            self.httpMethod = "HEAD"
            break

        case .link:
            self.httpMethod = "LINK"
            break

        case .options:
            self.httpMethod = "OPTIONS"
            break

        case .patch:
            self.httpMethod = "PATCH"
            break

        case .post:
            self.httpMethod = "POST"
            break

        case .purge:
            self.httpMethod = "PURGE"
            break

        case .put:
            self.httpMethod = "PUT"
            break

        case .unlink:
            self.httpMethod = "UNLINK"
            break


        default:
            self.httpMethod = "GET"
            break
        }
    }
}
